libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -2,43 +2,102 @@
+} from './cart.error';
+import { CartManager } from './cart.manager';
+import { ResultCart, TaxAddress } from './cart.types';
+import { handleEligibilityStatusMap } from './cart.utils';
 
 @Injectable()
 export class CheckoutService {
   constructor(
     private stripeClient: StripeClient,
     private stripeManager: StripeManager,
     private paypalCustomerManager: PaypalCustomerManager,
-    private paypalManager: PayPalManager
+    private paypalManager: PayPalManager,
+    private cartManager: CartManager,
+    private eligibilityService: EligibilityService
   ) {}
 
   async prePaySteps(cart: ResultCart) {
-    // TODO:
-    // - If uid not present, create user stub account and update cart
-    // - Check if customer exists, create via stripe manager and update cart
-    // - Make sure shipping address of customer is updated with cart taxAddress
-    // - Check if customer is eligible for product in cart
-    // - Check if amount to be charged matches amount in cart
-    // - Cancel incomplete subscriptions to price user is subscribing to
-
-    const { stripeCustomerId } = cart;
-    if (!stripeCustomerId) {
-      // TODO: this is a placeholder and won't be necessary once the above todo items are done
-      throw new Error("Stripe customer doesn't exist");
+    let customer;
+
+    // if stripeCustomerId does not exist:
+    // create stub via StripeClient and update cart
+    if (!cart.uid || !cart.stripeCustomerId) {
+      const customerStub = await this.stripeClient.customersCreate();
+
+      await this.cartManager.updateFreshCart(cart.id, cart.version, {
+        stripeCustomerId: customerStub.id,
+      });
+
+      customer = await this.stripeManager.fetchActiveCustomer(customerStub.id);
+    } else {
+      customer = await this.stripeManager.fetchActiveCustomer(
+        cart.stripeCustomerId
+      );
     }
-    const customer = await this.stripeManager.fetchActiveCustomer(
-      stripeCustomerId
+
+    // update customer shipping address with cart taxAddress
+    const taxAddress = cart.taxAddress as any as TaxAddress;
+
+    await this.stripeManager.updateCustomer(customer.id, {
+      address: {
+        country: taxAddress.countryCode,
+        postal_code: taxAddress.postalCode,
+      },
+    });
+
+    // validate customer is eligible for product via eligibility service
+    // throws cart eligibility mismatch error if no match found
+    const eligibility = await this.eligibilityService.checkEligibility(
+      cart.interval,
+      cart.offeringConfigId,
+      customer.id
     );
 
+    const cartEligibilityStatus = handleEligibilityStatusMap[eligibility];
+
+    if (cartEligibilityStatus !== cart.eligibilityStatus) {
+      throw new CartEligibilityMismatchError(cart.id);
+    }
+
+    // re-validate total amount against upcoming invoice via StripeClient
+    // throws cart total mismatch error if no match found
+    const upcomingInvoice = await this.stripeClient.invoicesRetrieveUpcoming();
```

Comment:
This should be retrieving the `upcomingInvoice` for a specific Price ID. 

The Price ID should be resolved from querying the CMS, using the cart.offeringConfigId + cart.interval. This work is currently being completed in [FXA-8903](https://mozilla-hub.atlassian.net/browse/FXA-8903). Once that PR lands you should be able to use the same logic to get the Price ID.